pkg/util/log/log.go

```diff
@@ -311,118 +339,138 @@ func formatError(v ...interface{}) error {
 	return errors.New(msg)
 }
 
-//Trace logs at the trace level
+// Trace logs at the trace level
 func Trace(v ...interface{}) {
 	if logger != nil && logger.inner != nil && logger.shouldLog(seelog.TraceLvl) {
 		s := buildLogEntry(v...)
 		logger.trace(logger.scrub(s))
+	} else if bufferLogsBeforeInit && (logger == nil || logger.inner == nil) {
+		addLogToBuffer(func() { Trace(v...) })
 	}
 }
 
-//Debug logs at the debug level
+// Debug logs at the debug level
 func Debug(v ...interface{}) {
 	if logger != nil && logger.inner != nil && logger.shouldLog(seelog.DebugLvl) {
 		s := buildLogEntry(v...)
 		logger.debug(logger.scrub(s))
+	} else if bufferLogsBeforeInit && (logger == nil || logger.inner == nil) {
+		addLogToBuffer(func() { Debug(v...) })
 	}
 }
 
-//Info logs at the info level
+// Info logs at the info level
 func Info(v ...interface{}) {
 	if logger != nil && logger.inner != nil && logger.shouldLog(seelog.InfoLvl) {
 		s := buildLogEntry(v...)
 		logger.info(logger.scrub(s))
+	} else if bufferLogsBeforeInit && (logger == nil || logger.inner == nil) {
+		addLogToBuffer(func() { Info(v...) })
 	}
 }
 
-//Warn logs at the warn level and returns an error containing the formated log message
+// Warn logs at the warn level and returns an error containing the formated log message
 func Warn(v ...interface{}) error {
 	if logger != nil && logger.inner != nil && logger.shouldLog(seelog.WarnLvl) {
 		s := buildLogEntry(v...)
 		return logger.warn(logger.scrub(s))
+	} else if bufferLogsBeforeInit && (logger == nil || logger.inner == nil) {
+		addLogToBuffer(func() { Warn(v...) })
 	}
 	return formatError(v...)
 }
 
-//Error logs at the error level and returns an error containing the formated log message
+// Error logs at the error level and returns an error containing the formated log message
 func Error(v ...interface{}) error {
 	if logger != nil && logger.inner != nil && logger.shouldLog(seelog.ErrorLvl) {
 		s := buildLogEntry(v...)
 		return logger.error(logger.scrub(s))
 	}
+	// We don't buffer error logs but directly print them on Stderr
```

Comment:
I think it make sense to both buffer them and print them to `stderr`. 
Since we can't predict if the error is going to stop the agent or not we still need to output to `stderr` but if we get to the point where the agent starts the logger it might be confusing / we might miss something if the error is not in the log file.

Reply:
➕ (and also applies to `critical` messages too, even though theoretically they are always supposed to be used right before a shutdown)

Reply:
Good point. I'm not sure what to do here since it would also print some error twice. I guess having twice an error message is better than not having it in the log file ?